internal/upload/gcp/gcp.go

```diff
@@ -0,0 +1,417 @@
+//                        ubuntu-2004, windows-10-x64-byol, windows-10-x86-byol,
+//                        windows-2008r2, windows-2008r2-byol, windows-2012, windows-2012-byol,
+//                        windows-2012r2, windows-2012r2-byol, windows-2016, windows-2016-byol,
+//                        windows-2019, windows-2019-byol, windows-7-x64-byol,
+//                        windows-7-x86-byol, windows-8-x64-byol, windows-8-x86-byol
+// region - A valid region where the resulting image should be located. If empty,
+//          the multi-region location closest to the source is chosen automatically.
+//          See: https://cloud.google.com/storage/docs/locations
+//
+// Uses:
+//	- Cloud Build API
+func (g *GCP) Import(bucket, object, imageName, os, region string) error {
+	ctx := context.Background()
+	cloudbuildClient, err := cloudbuild.NewClient(ctx, option.WithCredentials(g.creds))
+	if err != nil {
+		return fmt.Errorf("failed to get Cloud Build client: %v", err)
+	}
+	defer cloudbuildClient.Close()
+
+	buildStepArgs := []string{
+		fmt.Sprintf("-source_file=gs://%s/%s", bucket, object),
+		fmt.Sprintf("-image_name=%s", imageName),
+		"-timeout=7000s",
+		"-client_id=api",
+	}
+	if region != "" {
+		buildStepArgs = append(buildStepArgs, fmt.Sprintf("-storage_location=%s", region))
+	}
+	if os != "" {
+		buildStepArgs = append(buildStepArgs, fmt.Sprintf("-os=%s", os))
+	} else {
+		// This effectively marks the image as non-bootable for the import process,
+		// but it has no effect on the later use or booting in Compute Engine other
+		// than the GCP guest tools not being installed.
+		buildStepArgs = append(buildStepArgs, "-data_disk")
+	}
+
+	imageBuild := &cloudbuildpb.Build{
+		Steps: []*cloudbuildpb.BuildStep{{
+			Name: "gcr.io/compute-image-tools/gce_vm_image_import:release",
+			Args: buildStepArgs,
+		}},
+		Tags: []string{
+			"gce-daisy",
+			"gce-daisy-image-import",
+		},
+		Timeout: durationpb.New(time.Second * 7200),
+	}
+
+	createBuildReq := &cloudbuildpb.CreateBuildRequest{
+		ProjectId: g.creds.ProjectID,
+		Build:     imageBuild,
+	}
+	fmt.Printf("[GCP] 📥 Importing image into Compute Node as '%s'\n", imageName)
+	resp, err := cloudbuildClient.CreateBuild(ctx, createBuildReq)
+	if err != nil {
+		return fmt.Errorf("failed to create image import build job: %v", err)
+	}
+
+	// Get the returned Build struct
+	buildOpMetadata := &cloudbuildpb.BuildOperationMetadata{}
+	if err := ptypes.UnmarshalAny(resp.Metadata, buildOpMetadata); err != nil {
+		return err
+	}
+	imageBuild = buildOpMetadata.Build
+
+	fmt.Printf("[GCP] 📜 Image import log URL: %s\n", imageBuild.LogUrl)
+	fmt.Printf("[GCP] 🤔 Image import build status: %+v\n", imageBuild.Status)
+
+	getBuldReq := &cloudbuildpb.GetBuildRequest{
+		ProjectId: imageBuild.ProjectId,
+		Id:        imageBuild.Id,
+	}
+
+	// Wait for the build to finish
+	fmt.Println("[GCP] 🥱 Waiting for the image import to finish")
+	for {
```

Comment:
In general I think we should allow things to be cancelled by the caller. Not a fan of timeouts, except as a very last resort at the highest level (usually I'd say the user getting bored and clicking "cancel" is timeout good enough).
Might want to do the cancel logic in a follow-up though, seems like more invasive surgery would be needed.

Reply:
Agreed!

Reply:
Yeah, we probably should. I'm open to suggestions. The GCP build job has a timeout set to 7200 sec (2 hours) [on line 190](https://github.com/osbuild/osbuild-composer/pull/1245/files/cc4759b408a16325f4aeef7b9cc3efb4ef1714e2#diff-fa36b8cedd24680d43cde61fbdcd8681c6ccf4ebc452e394c34e692d575af6d3R190), so it will eventually fail. I could either expose this timeout or pass a context as you suggested. The magic value of 7200 sec is from GCP API example...

I didn't do it for the first iteration, because none of the internal cloud provider APIs allow to set timeout while they are still blocking.

Reply:
I'd keep it like this for now. Then go through in a follow-up and make sure all of  this (for all the providers) is cancelable.